File: python/session_subscribe.py
import asyncio
import json
import time

from config import Config
from pxgrid import PxgridControl
from websockets import ConnectionClosed
from ws_stomp import WebSocketStomp


async def subscribe_loop(config, secret, ws_url, topic):
    ws = WebSocketStomp(ws_url, config.get_node_name(),
                        secret, config.get_ssl_context())
    await ws.connect()
    await ws.stomp_connect(pubsub_node_name)
    await ws.stomp_subscribe(topic, config.get_filter())
    print("Press <Ctrl-C> to disconnect")
    while True:
        try:
            message = await ws.stomp_read_message()
            message_json = json.loads(message)
            print("message=" + json.dumps(message_json))
        except ConnectionClosed:
            print('Websocket connection closed')
            break
        except asyncio.CancelledError:
            await ws.stomp_disconnect('123')
            # wait for receipt
            await asyncio.sleep(3)
            await ws.disconnect()
            print("Websocket connection disconnected")
            break


if __name__ == '__main__':
    config = Config()
    pxgrid = PxgridControl(config=config)

    while pxgrid.account_activate()['accountState'] != 'ENABLED':
        time.sleep(60)

    # lookup for session service
    service_lookup_response = pxgrid.service_lookup('com.cisco.ise.session')
    service = service_lookup_response['services'][0]
    pubsub_service_name = service['properties']['wsPubsubService']
    topic = service['properties']['sessionTopic']

    # lookup for pubsub service
    service_lookup_response = pxgrid.service_lookup(pubsub_service_name)
    pubsub_service = service_lookup_response['services'][0]
    pubsub_node_name = pubsub_service['nodeName']
    secret = pxgrid.get_access_secret(pubsub_node_name)['secret']
    ws_url = pubsub_service['properties']['wsUrl']

    asyncio.run(subscribe_loop(config, secret, ws_url, topic))

File: golang/session_subscribe.go
package main

import (
	"log"
	"os"
	"os/signal"
	"sync"
	"syscall"
	"time"
)

func dataPrinter(dataChan <-chan *EndpointData) {
	for data := range dataChan {
		if data.Err == nil {
			log.Println("Message=" + string(data.Content))
		} else {
			log.Println(data.Err)
		}
	}
}

func main() {
	config := NewConfig()
	control, err := NewControl(config)
	if err != nil {
		log.Fatal(err)
	}

	// AccountActivate
	for {
		res, err := control.AccountActivate()
		if err != nil {
			log.Fatal(err)
		}
		if res.AccountState == "ENABLED" {
			break
		}
		time.Sleep(30 * time.Second)
	}

	// pxGrid ServiceLookup for Session Directory
	services, err := control.ServiceLookup(config.service)
	if err != nil {
		log.Fatal(err)
	} else if len(services) == 0 {
		log.Fatal("Service unavailable")
	}

	// Use first service
	wsPubsubService := services[0].Properties["wsPubsubService"]
	topic := services[0].Properties[config.topic]
	log.Println("wsPubsubService=", wsPubsubService, "topic=", topic)

	// pxGrid ServiceLookup for pubsub service
	pubsubServices, err := control.ServiceLookup(wsPubsubService)
	if err != nil {
		log.Fatal(err)
	} else if len(pubsubServices) == 0 {
		log.Fatal("Pubsub service unavailable")
	}

	// Use first pubsub service
	pubsubService := pubsubServices[0]
	pubsubNodeName := pubsubService.NodeName
	wsUrl := pubsubService.Properties["wsUrl"]
	log.Println("wsUrl=", wsUrl)

	// pxGrid AccessSecret with the pubsub node
	secret, err := control.GetAccessSecret(pubsubNodeName)
	if err != nil {
		log.Fatal(err)
	}

	// Setup WebSocket
	endpoint, err := NewEndpoint(config)
	if err != nil {
		log.Fatal(err)
	}

	err = endpoint.Connect(wsUrl, config.nodeName, secret)
	if err != nil {
		log.Fatal(err)
	}

	err = endpoint.Subscribe(topic, config.filter)
	if err != nil {
		log.Fatal(err)
	}

	dataChan := make(chan *EndpointData)
	go dataPrinter(dataChan)
	wg := sync.WaitGroup{}

	go func() {
		wg.Add(1)
		endpoint.Listener(dataChan)
		wg.Done()
	}()

	// Setup abort channel
	log.Println("Press <Ctrl-c> to disconnect...")
	abort := make(chan os.Signal, 1)
	signal.Notify(abort, os.Interrupt, syscall.SIGTERM)
	<-abort

	// Cleanup
	log.Printf("Disconnecting websocket connection...")
	endpoint.Disconnect()

	wg.Wait()
	close(dataChan)
}
